tracer/src/Datadog.Trace/Ci/Agent/CIAgentlessWriter.cs

```diff
@@ -0,0 +1,217 @@
+// <copyright file="CIAgentlessWriter.cs" company="Datadog">
+// Unless explicitly stated otherwise all files in this repository are licensed under the Apache 2 License.
+// This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2017 Datadog, Inc.
+// </copyright>
+
+using System;
+using System.Collections.Concurrent;
+using System.Threading;
+using System.Threading.Tasks;
+using Datadog.Trace.Ci.Agent.Payloads;
+using Datadog.Trace.Ci.EventModel;
+using Datadog.Trace.Configuration;
+using Datadog.Trace.Logging;
+using Datadog.Trace.Sampling;
+
+namespace Datadog.Trace.Ci.Agent
+{
+    /// <summary>
+    /// CI Visibility Agentless Writer
+    /// </summary>
+    internal sealed class CIAgentlessWriter : IEventWriter
+    {
+        private const int BatchInterval = 1000;
+        private const int MaxItemsInQueue = 2500;
+
+        private static readonly IDatadogLogger Log = DatadogLogging.GetLoggerFor<CIAgentlessWriter>();
+
+        private readonly BlockingCollection<IEvent> _eventQueue;
+        private readonly TaskCompletionSource<bool> _flushTaskCompletionSource;
+        private readonly Task _periodicFlush;
+        private readonly AutoResetEvent _flushDelayEvent;
+
+        private readonly EventsPayload _ciTestCycleBuffer;
+
+        private readonly ICIAgentlessWriterSender _sender;
+
+        public CIAgentlessWriter(ImmutableTracerSettings settings, ISampler sampler, ICIAgentlessWriterSender sender)
+        {
+            _eventQueue = new BlockingCollection<IEvent>(MaxItemsInQueue);
+            _flushTaskCompletionSource = new TaskCompletionSource<bool>(TaskCreationOptions.RunContinuationsAsynchronously);
+            _flushDelayEvent = new AutoResetEvent(false);
+
+            _ciTestCycleBuffer = new CITestCyclePayload();
+
+            _sender = sender;
+
+            _periodicFlush = Task.Factory.StartNew(InternalFlushEventsAsync, this, TaskCreationOptions.LongRunning);
+            _periodicFlush.ContinueWith(t => Log.Error(t.Exception, "Error in sending ciapp events"), TaskContinuationOptions.OnlyOnFaulted);
+
+            Log.Information("CIAgentlessWriter Initialized.");
+        }
+
+        public void WriteEvent(IEvent @event)
+        {
+            if (_eventQueue.IsAddingCompleted)
+            {
+                return;
+            }
+
+            try
+            {
+                _eventQueue.Add(@event);
```

Comment:
Are we ok with this blocking indefinitely? If there's an issue writing to the endpoint (e.g. the service is down, I think the blocking collection could fill up, and then this will block). What are the implications of that (I haven't followed it through yet). Worst case, it blocks user code writing a trace, causing test hangs. Best case, it only affects the writer thread, and we cancel the add on app close? That latter approach would prob need `TryAdd` with cancellation though right?

Reply:
Yes for CI Visibility we rather prefer to block (stop the world situation) until the payload is sent or dropped due timeout, after retrying several times. We cannot affort loosing data. I have test it with an invalid endpoint, or sending the wrong content type, and receiving backend error, and definitely slows down the test execution after reaching the max number of tests events (MaxItemsInQueue = 2500). Also, yes when the testhost is about to close we call `FlushAndCloseAsync` and that sets the `CompleteAdding` flag for future `WriteEvent` calls so later calls will not get affected. I have to think about adding a CancellationToken in this scenario too but, I don't want to end up in a situation where because the endpoint is slow then we cancel prematurely.

Reply:
> Yes for CI Visibility we rather prefer to block

Interesting, I would _absolutely not want this_. Otherwise, Datadog CI endpoint going down could kill all CI and the ability for us to release (we could easily exceed timeouts, and have no easy way of getting CI to pass). 😬

Reply:
I think we can aliviate this by reducing the Sender timeouts. So if the CI endpoint is down we drop the entire payload faster and unblock the test process. WDYT?

I will bring this to the next CIApp guild meeting to discuss this scenario. Thanks for the feedback.

Reply:
> I think we can aliviate this by reducing the Sender timeouts

Yes, I think this is something we should do on _all_ our HttpClients tbh, particularly the agentless ones like Logs/Telemetry/CIapp, but maybe also for the "agent" ones too; as the defaults are typically to high. I don't think it should be a blocker, but definitely something to consider